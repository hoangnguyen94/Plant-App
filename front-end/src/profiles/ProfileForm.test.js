import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../auth/UserContext";
import ProfileForm from "./ProfileForm";
import PlantApi from "../api/api";
import '@testing-library/jest-dom/extend-expect';
import "mutationobserver-shim";

// Mock the UserContext
const mockUserContextValue = {
  currentUser: {
    username: "testuser",
    firstName: "Test",
    lastName: "Tester",
    email: "test@example.com"
  },
  setCurrentUser: jest.fn()
};

// Mock the PlantApi.saveProfile method
jest.mock("../api/api", () => ({
  saveProfile: jest.fn(() => ({
    username: "testuser",
    firstName: "Updated First Name",
    lastName: "Updated Last Name",
    email: "updated@example.com"
  }))
}));

describe("ProfileForm", () => {
  it("renders profile form with initial values", () => {
    const { getByLabelText, container } = render(
      <MemoryRouter>
        <UserContext.Provider value={mockUserContextValue}>
          <ProfileForm />
        </UserContext.Provider>
      </MemoryRouter>
	);
	  
    // Assert that the form fields have the initial values
    expect(getByLabelText("Username")).toHaveValue("testuser");
    expect(getByLabelText("First Name")).toHaveValue("Test");
    expect(getByLabelText("Last Name")).toHaveValue("Tester");
    expect(getByLabelText("Email")).toHaveValue("test@example.com");
    expect(getByLabelText("Confirm password to make changes:")).toHaveValue("");

    // Assert snapshot
    expect(container.firstChild).toMatchSnapshot();
  });

  it("handles form submission and updates user information", async () => {
    const { getByLabelText, getByText, queryByText, container } = render(
      <MemoryRouter>
        <UserContext.Provider value={mockUserContextValue}>
          <ProfileForm />
        </UserContext.Provider>
      </MemoryRouter>
    );

    // Simulate changes in form fields
    fireEvent.change(getByLabelText("First Name"), {
      target: { value: "Updated First Name" }
    });
    fireEvent.change(getByLabelText("Last Name"), {
      target: { value: "Updated Last Name" }
    });
    fireEvent.change(getByLabelText("Email"), {
      target: { value: "updated@example.com" }
    });
    fireEvent.change(getByLabelText("Confirm password to make changes:"), {
      target: { value: "password123" }
    });

    // Simulate form submission
    fireEvent.click(getByText("Save Changes"));

    // Assert that the PlantApi.saveProfile method is called with the correct arguments
    expect(PlantApi.saveProfile).toHaveBeenCalledWith(
      "testuser",
      {
        firstName: "Updated First Name",
        lastName: "Updated Last Name",
        email: "updated@example.com",
        password: "password123"
      }
    );
    
    mockUserContextValue.setCurrentUser({
      username: "testuser",
      firstName: "Updated First Name",
      lastName: "Updated Last Name",
      email: "updated@example.com"
    });
    // Assert that the setCurrentUser method is called with the updated user information
    expect(mockUserContextValue.setCurrentUser).toHaveBeenCalledWith({
      username: "testuser",
      firstName: "Updated First Name",
      lastName: "Updated Last Name",
      email: "updated@example.com"
    });
    // Simulate form submission
    fireEvent.click(getByText("Save Changes"));

    // Wait for the success message to appear
    await waitFor(() =>
      expect(queryByText("Updated successfully.")).toBeInTheDocument()
    );
    // Assert snapshot
    expect(container.firstChild).toMatchSnapshot();
  });
});
