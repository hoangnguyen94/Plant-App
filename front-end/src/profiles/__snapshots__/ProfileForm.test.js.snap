// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm handles form submission and updates user information 1`] = `
<div
  class="col-md-6 col-lg-4 offset-md-3 offset-lg-4"
>
  <h3>
    Profile
  </h3>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <form>
        <div
          class="form-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            class="form-control-plaintext"
            id="username"
            readonly=""
            value="testuser"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            value="Updated First Name"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            value="Updated Last Name"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            value="updated@example.com"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Confirm password to make changes:
          </label>
          <input
            class="form-control"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="alert alert-success"
          role="alert"
        >
          <p
            class="mb-0 small"
          >
            Updated successfully.
          </p>
        </div>
        <button
          class="btn btn-primary btn-block mt-4"
        >
          Save Changes
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`ProfileForm renders profile form with initial values 1`] = `
<div
  class="col-md-6 col-lg-4 offset-md-3 offset-lg-4"
>
  <h3>
    Profile
  </h3>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <form>
        <div
          class="form-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            class="form-control-plaintext"
            id="username"
            readonly=""
            value="testuser"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            value="Test"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            value="Tester"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            value="test@example.com"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Confirm password to make changes:
          </label>
          <input
            class="form-control"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <button
          class="btn btn-primary btn-block mt-4"
        >
          Save Changes
        </button>
      </form>
    </div>
  </div>
</div>
`;
